libbeat/processors/ratelimit/token_bucket.go

```diff
@@ -35,10 +35,36 @@ func init() {
 }
 
 type bucket struct {
+	mu sync.Mutex
+
 	tokens        float64
 	lastReplenish time.Time
 }
 
+func (b *bucket) withdraw() bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
```

Comment:
@mauri870 did you try explicitly unlocking before returning instead of using defer?

I remember defer statements having a cost that is somehow considerable. I recall improvements there but I'm not completely sure how considerable it is in the Go versions we're using in Beats.

Because those functions are so small and it is in hot path, I believe it is worth comparing the implementations.

Reply:
That used to be a problem, but since 1.14 defer incurs almost zero overhead. I just tested in this particular scenario and the cost is 1ns. https://go.dev/doc/go1.14#runtime

Reply:
I think this can be made faster by using x/time/rate or a lock-free algorithm. I can follow up on this with another PR. For now I just want to get this fixed.

Reply:
Actually, I noticed that there is an allocation taking place on the hot path. Getting rid of it gives a 25% improvement.